src/transform/src/attribute/cardinality.rs

```diff
@@ -0,0 +1,646 @@
+                        Some(SymbolicExpression::symbolic(FactorizerVariable::Index(
+                            *col,
+                        )))
+                    } else {
+                        None
+                    }
+                }
+                _ => None,
+            }
+        };
+
+        match expr {
+            MirScalarExpr::Column(_)
+            | MirScalarExpr::Literal(_, _)
+            | MirScalarExpr::CallUnmaterializable(_) => SymExp::from(1.0),
+            MirScalarExpr::CallUnary { func, expr } => match func {
+                UnaryFunc::Not(_) => 1.0 - self.predicate(expr, indexed_columns),
+                UnaryFunc::IsTrue(_) | UnaryFunc::IsFalse(_) => SymExp::from(0.5),
+                UnaryFunc::IsNull(_) => {
+                    if let Some(icard) = index_cardinality(expr) {
+                        icard
+                    } else {
+                        SymExp::from(WORST_CASE_SELECTIVITY)
+                    }
+                }
+                _ => SymExp::from(WORST_CASE_SELECTIVITY),
+            },
+            MirScalarExpr::CallBinary { func, expr1, expr2 } => {
+                match func {
+                    BinaryFunc::Eq => match (index_cardinality(expr1), index_cardinality(expr2)) {
+                        (Some(icard1), Some(icard2)) => SymbolicExpression::max(icard1, icard2),
+                        (Some(icard), None) | (None, Some(icard)) => icard,
+                        (None, None) => SymExp::from(WORST_CASE_SELECTIVITY),
+                    },
+                    // 1.0 - the Eq case
+                    BinaryFunc::NotEq => match (index_cardinality(expr1), index_cardinality(expr2))
+                    {
+                        (Some(icard1), Some(icard2)) => {
+                            1.0 - SymbolicExpression::max(icard1, icard2)
+                        }
+                        (Some(icard), None) | (None, Some(icard)) => 1.0 - icard,
+                        (None, None) => SymExp::from(1.0 - WORST_CASE_SELECTIVITY),
+                    },
+                    BinaryFunc::Lt | BinaryFunc::Lte | BinaryFunc::Gt | BinaryFunc::Gte => {
+                        // if we have high/low key values and one of the columns is an index
+                        SymExp::from(0.33)
+                    }
+                    _ => SymExp::from(1.0), // TOOD(mgree): are there other interesting cases?
+                }
+            }
+            MirScalarExpr::CallVariadic { func, exprs } => match func {
+                VariadicFunc::And => {
+                    // can't use SymExp::product because it expects a vector of references :/
+                    let mut factor = SymExp::from(1.0);
+
+                    for expr in exprs {
+                        factor = factor * self.predicate(expr, indexed_columns);
+                    }
+
+                    factor
+                }
+                VariadicFunc::Or => {
```

Comment:
I'm slightly worried about the size of the symbolic expressions that would come out after this step because of the `F(expr1) * F(expr2)` factor.

Reply:
Fair! Right now we're not doing much simplification, which might help with that. Hash-consing the expressions would give significant storage savings, too. Added a TODO.